File: src/main/scala/com/martomate/tripaint/model/FloodFillSearcher.scala
package com.martomate.tripaint.model

import com.martomate.tripaint.model.coords.GlobalPixCoords

class FloodFillSearcher(colorLookup: ColorLookup) {
  def search(
      startPos: GlobalPixCoords,
      predicate: (GlobalPixCoords, Color) => Boolean
  ): Seq[GlobalPixCoords] = {
    val visited = collection.mutable.Set.empty[GlobalPixCoords]
    val result = collection.mutable.ArrayBuffer.empty[GlobalPixCoords]
    val q = collection.mutable.Queue(startPos)
    visited += startPos

    while (q.nonEmpty) {
      val p = q.dequeue()
      colorLookup.lookup(p) foreach { color =>
        if (predicate(p, color)) {
          result += p

          val newOnes = p.neighbours.filter(!visited(_))
          visited ++= newOnes
          q ++= newOnes
        }
      }
    }
    result.toSeq
  }
}

File: core/src/main/kotlin/tripaint/FloodFillSearcher.kt
package tripaint

import tripaint.color.Color
import tripaint.coords.GlobalPixCoords

class FloodFillSearcher(private val colorLookup: ColorLookup) {
    fun search(
        startPos: GlobalPixCoords,
        predicate: (GlobalPixCoords, Color) -> Boolean
    ): List<GlobalPixCoords> {
        val visited: MutableSet<GlobalPixCoords> = mutableSetOf()
        val result: MutableList<GlobalPixCoords> = mutableListOf()
        val q = ArrayDeque(listOf(startPos))
        visited += startPos

        while (!q.isEmpty()) {
            val p = q.removeFirst()
            val color = colorLookup.lookup(p)
            if (color != null) {
                if (predicate(p, color)) {
                    result += p

                    val newOnes = p.neighbours().filter { !visited.contains(it) }
                    visited.addAll(newOnes)
                    q.addAll(newOnes)
                }
            }
        }
        return result
    }
}
